File: ts/2022/day23.ts
import { main } from 'lib/advent';
import {
  Dir,
  findBounds,
  move,
  parseGrid,
  Point,
  PointMap,
  PointSet,
} from 'lib/coords';
import { iter } from 'lib/iter';
import { lines } from 'lib/util';

function parse(lines: string[]) {
  return new PointSet(
    iter(parseGrid(lines, (c) => c === '#'))
      .filter(([, isElf]) => isElf)
      .map(([p]) => p)
      .toArray()
  );
}

function neighbors({ x, y }: Point, elves: PointSet): Record<string, boolean> {
  return {
    N: elves.has({ x, y: y - 1 }),
    E: elves.has({ x: x + 1, y }),
    S: elves.has({ x, y: y + 1 }),
    W: elves.has({ x: x - 1, y }),
    NW: elves.has({ x: x - 1, y: y - 1 }),
    SW: elves.has({ x: x - 1, y: y + 1 }),
    SE: elves.has({ x: x + 1, y: y + 1 }),
    NE: elves.has({ x: x + 1, y: y - 1 }),
  };
}

function part1(elves: PointSet, rounds = 10): number {
  const proposals = [
    { test: (n) => !n.N && !n.NE && !n.NW, dir: Dir.Up },
    { test: (n) => !n.S && !n.SE && !n.SW, dir: Dir.Down },
    { test: (n) => !n.W && !n.NW && !n.SW, dir: Dir.Left },
    { test: (n) => !n.E && !n.NE && !n.SE, dir: Dir.Right },
  ];

  for (let i = 0; i < rounds; ++i) {
    // Propose moves
    const moves = new PointMap<Point>();
    for (const elf of elves) {
      const n = neighbors(elf, elves);
      if (Object.values(n).every((v) => !v)) continue;
      else {
        for (const p of proposals) {
          if (p.test(n)) {
            moves.set(elf, move(elf, p.dir));
            break;
          }
        }
      }
    }

    // Check moves
    for (const next of moves.values()) {
      const samePoint = iter(moves).filter(
        ([, to]) => to.x === next.x && to.y === next.y
      );
      if (samePoint.size > 1) {
        samePoint.forEach(([from]) => moves.delete(from));
      }
    }

    if (moves.size === 0) return i + 1;

    // Move
    for (const [cur, next] of moves) {
      elves.delete(cur);
      elves.add(next);
    }

    // Rotate proposals
    proposals.push(proposals.shift());
  }

  const { min, max } = findBounds(elves);
  return (max.x - min.x + 1) * (max.y - min.y + 1) - elves.size;
}

// TODO: cleanup
// TODO: optimize p2: 48s
main(
  (s) => part1(new PointSet(parse(lines(s)))),
  (s) => part1(parse(lines(s)), Infinity)
);

File: py/2022/day23.py
from collections import deque
from itertools import count

from aoc import main
from aoc.coords import Dir8 as Dir
from aoc.coords import Point, addp, find_bounds


def parse(s: str) -> set[Point]:
    elves = set()
    for y, line in enumerate(s.splitlines()):
        for x, c in enumerate(line):
            if c == "#":
                elves.add((x, y))
    return elves


def elf_life(s: str, rounds: int | None = None) -> int:
    elves = parse(s)
    rules = deque(
        [
            (Dir.N, set([Dir.N, Dir.NE, Dir.NW])),
            (Dir.S, set([Dir.S, Dir.SE, Dir.SW])),
            (Dir.W, set([Dir.W, Dir.NW, Dir.SW])),
            (Dir.E, set([Dir.E, Dir.NE, Dir.SE])),
        ]
    )

    dirs = list(Dir)  # optimization
    iterator = count(1) if rounds is None else range(rounds)
    for round in iterator:
        moves = 0
        current_elves = elves.copy()
        for elf in current_elves:
            neighbors = {d for d in dirs if addp(elf, d) in current_elves}
            if not neighbors:
                continue
            for dir, check in rules:
                if not neighbors & check:
                    next = addp(elf, dir)
                    if next in elves:
                        # Only two elves in opposing directions can conflict.
                        moves -= 1
                        elves.remove(next)
                        elves.add(addp(next, dir))
                    else:
                        moves += 1
                        elves.add(next)
                        elves.remove(elf)
                    break

        if moves == 0:
            return round
        rules.rotate(-1)

    (ax, ay), (bx, by) = find_bounds(elves)
    rect = (bx + 1 - ax) * (by + 1 - ay)
    return rect - len(elves)


# TODO optimize part 2; currently 1.91s
if __name__ == "__main__":
    main(
        lambda s: elf_life(s, 10),
        lambda s: elf_life(s),
    )
